File: apps/api/app/services/compose.py
"""Ingredient Composer math — Python port of packages/nutrition-core/src/compose.ts.

Keep behavior identical between the TS and Python implementations so the live
frontend preview matches the backend-authoritative total.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from app.models.food import Unit

Cooking = Literal["raw", "boiled", "fried", "baked", "grilled"]

# Volume → ml approximations when only a volume unit is known.
VOLUME_TO_ML: dict[Unit, float] = {
    Unit.cup: 200.0,  # piyola
    Unit.tbsp: 15.0,  # osh qoshiq
    Unit.tsp: 5.0,  # chay qoshiq
}

# Nutrient profile for sunflower oil (per 100g).
OIL_PER_100G: dict[str, float] = {"kcal": 884.0, "fat": 100.0, "vit_e": 41.08}
OIL_DENSITY = 0.92  # g/ml


@dataclass(frozen=True)
class UnitContext:
    grams_per_piece: float | None = None
    grams_per_serving: float | None = None
    density_g_per_ml: float = 1.0


@dataclass(frozen=True)
class ComposerIngredientIn:
    food_id: int
    quantity: float
    unit: Unit
    per_100g: dict[str, float]  # fetched from DB
    ctx: UnitContext = field(default_factory=UnitContext)
    cooking: Cooking | None = None
    added_oil_ml: float | None = None


@dataclass
class ComposedTotals:
    total_grams: float = 0.0
    kcal: float = 0.0
    protein: float = 0.0
    carbs: float = 0.0
    fat: float = 0.0
    micronutrients: dict[str, float] = field(default_factory=dict)


def to_grams(qty: float, unit: Unit, ctx: UnitContext) -> float:
    """Convert any Unit + quantity to grams for nutrient scaling."""
    density = ctx.density_g_per_ml
    if unit == Unit.g:
        return qty
    if unit == Unit.ml:
        return qty * density
    if unit == Unit.piece:
        return qty * (ctx.grams_per_piece or 100.0)
    if unit == Unit.serving:
        return qty * (ctx.grams_per_serving or 100.0)
    if unit in VOLUME_TO_ML:
        return qty * VOLUME_TO_ML[unit] * density
    raise ValueError(f"Unknown unit: {unit}")


def scale_nutrient(per_100g: float, grams: float) -> float:
    return (per_100g * grams) / 100.0


def compose(ingredients: list[ComposerIngredientIn]) -> ComposedTotals:
    totals = ComposedTotals()

    for ing in ingredients:
        grams = to_grams(ing.quantity, ing.unit, ing.ctx)
        totals.total_grams += grams
        for key, per100 in ing.per_100g.items():
            _add(totals, key, scale_nutrient(per100, grams))

        # Fried → add oil by volume
        if ing.cooking == "fried" and ing.added_oil_ml:
            oil_g = ing.added_oil_ml * OIL_DENSITY
            totals.total_grams += oil_g
            for key, per100 in OIL_PER_100G.items():
                _add(totals, key, scale_nutrient(per100, oil_g))

    # Round for display
    totals.kcal = round(totals.kcal)
    totals.protein = round(totals.protein, 1)
    totals.carbs = round(totals.carbs, 1)
    totals.fat = round(totals.fat, 1)
    totals.total_grams = round(totals.total_grams, 1)
    for k in list(totals.micronutrients):
        totals.micronutrients[k] = round(totals.micronutrients[k], 2)

    return totals


def _add(totals: ComposedTotals, key: str, value: float) -> None:
    if key == "kcal":
        totals.kcal += value
    elif key == "protein":
        totals.protein += value
    elif key == "carbs":
        totals.carbs += value
    elif key == "fat":
        totals.fat += value
    else:
        totals.micronutrients[key] = totals.micronutrients.get(key, 0.0) + value

File: packages/nutrition-core/src/compose.ts
/**
 * Ingredient Composer math — sum up nutrients for a user-built meal.
 *
 * This mirrors the backend `/api/ingredients/compose` endpoint so the UI can
 * show a *live preview* of totals before hitting the network.
 */

import { toGrams, scaleNutrient, type Unit, type UnitContext } from './units';

export interface NutrientMap {
  kcal: number;
  protein: number;
  carbs: number;
  fat: number;
  [key: string]: number;
}

export interface ComposerIngredient {
  id: string | number;
  name: string;
  quantity: number;
  unit: Unit;
  /** nutrients per 100g — from Food DB */
  per100g: Partial<NutrientMap>;
  /** unit context */
  ctx?: UnitContext;
  /** cooking method affects oil added */
  cooking?: 'raw' | 'boiled' | 'fried' | 'baked' | 'grilled';
  addedOilMl?: number;
}

export interface ComposedTotals {
  totalGrams: number;
  kcal: number;
  protein: number;
  carbs: number;
  fat: number;
  micronutrients: Record<string, number>;
}

/** Sunflower oil nutrient profile — per 100g = per 100ml × density (0.92). */
const OIL_PER_100G = { kcal: 884, fat: 100, vit_e: 41.08 };
const OIL_DENSITY = 0.92;

export function compose(ingredients: ComposerIngredient[]): ComposedTotals {
  const totals: ComposedTotals = {
    totalGrams: 0,
    kcal: 0,
    protein: 0,
    carbs: 0,
    fat: 0,
    micronutrients: {},
  };

  for (const ing of ingredients) {
    const grams = toGrams(ing.quantity, ing.unit, ing.ctx);
    totals.totalGrams += grams;
    for (const [key, per100] of Object.entries(ing.per100g)) {
      if (per100 === undefined) continue;
      const v = scaleNutrient(per100, grams);
      addMacroOrMicro(totals, key, v);
    }
    // Fried → add oil by volume
    if (ing.cooking === 'fried' && ing.addedOilMl) {
      const oilG = ing.addedOilMl * OIL_DENSITY;
      totals.totalGrams += oilG;
      for (const [key, per100] of Object.entries(OIL_PER_100G)) {
        addMacroOrMicro(totals, key, scaleNutrient(per100, oilG));
      }
    }
  }

  // Round for display
  totals.kcal = Math.round(totals.kcal);
  totals.protein = round1(totals.protein);
  totals.carbs = round1(totals.carbs);
  totals.fat = round1(totals.fat);
  for (const k in totals.micronutrients) {
    const current = totals.micronutrients[k];
    if (current !== undefined) totals.micronutrients[k] = round1(current);
  }

  return totals;
}

function addMacroOrMicro(totals: ComposedTotals, key: string, value: number): void {
  if (key === 'kcal' || key === 'protein' || key === 'carbs' || key === 'fat') {
    totals[key] += value;
  } else {
    totals.micronutrients[key] = (totals.micronutrients[key] ?? 0) + value;
  }
}

function round1(n: number): number {
  return Math.round(n * 10) / 10;
}
